src/charm.py

```diff
@@ -0,0 +1,292 @@
+        self._amf_metrics_endpoint = MetricsEndpointProvider(self)
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="prometheus-exporter", port=PROMETHEUS_PORT),
+                ServicePort(name="sbi", port=SBI_PORT),
+                ServicePort(name="ngapp", port=NGAPP_PORT, protocol="SCTP"),
+                ServicePort(name="sctp-grpc", port=SCTP_GRPC_PORT),
+            ],
+        )
+        self._amf_database = DatabaseRequires(
+            self, relation_name="amf-database", database_name=AMF_DATABASE_NAME
+        )
+        self._default_database = DatabaseRequires(
+            self, relation_name="default-database", database_name=DEFAULT_DATABASE_NAME
+        )
+
+        self.framework.observe(
+            self.on.default_database_relation_joined,
+            self._on_amf_pebble_ready,
+        )
+        self.framework.observe(
+            self.on.amf_database_relation_joined,
+            self._on_amf_pebble_ready,
+        )
+        self.framework.observe(
+            self._default_database.on.database_created, self._on_amf_pebble_ready
+        )
+        self.framework.observe(self._amf_database.on.database_created, self._on_amf_pebble_ready)
+        self.framework.observe(self.on.amf_pebble_ready, self._on_amf_pebble_ready)
+        self.framework.observe(self._nrf_requires.on.nrf_available, self._on_amf_pebble_ready)
+
+    def _on_amf_pebble_ready(
+        self,
+        event: EventBase,
+    ) -> None:
+        """Handle pebble ready event for AMF container.
+
+        Args:
+            event (PebbleReadyEvent, DatabaseCreatedEvent, NRFAvailableEvent): Juju event
+        """
+        if not self._amf_container.can_connect():
+            self.unit.status = MaintenanceStatus("Waiting for service to start")
+            event.defer()
+            return
+        for relation in ["fiveg_nrf", "amf-database", "default-database"]:
+            if not self._relation_created(relation):
+                self.unit.status = BlockedStatus(f"Waiting for {relation} relation")
+                return
+        if not self._default_database_is_available:
+            self.unit.status = WaitingStatus("Waiting for the default database to be available")
+            event.defer()
+            return
+        if not self._amf_database_is_available:
+            self.unit.status = WaitingStatus("Waiting for the amf database to be available")
+            event.defer()
+            return
+        if not self._default_database_info:
+            self.unit.status = WaitingStatus("Waiting for default database info to be available")
+            event.defer()
+            return
+        if not self._nrf_data_is_available:
+            self.unit.status = WaitingStatus("Waiting for NRF data to be available")
+            event.defer()
+            return
+        if not self._config_file_is_pushed:
```

Comment:
We'll also probably want to validate that:
- If the file is already created: the content matches what we expect it to be. This will be useful whenever any of the file parameter change (ex. NRF url).

Reply:
[Here's](https://github.com/canonical/sdcore-upf-operator/pull/1/files#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR195) an example of this.

Reply:
Done.